### PR TITLE
Fix build issues for idf 4.2 (Support ESP32-S2)

### PR DESCRIPTION
**Pull request by misery**
https://github.com/espressif/arduino-esp32/commit/ecc96060dab2fb954f29d36027b2cb21ba3dbfbc

Tested with:
```
esphome:
  name: esp32s2demo
  platform: ESP32
  board: nodemcu-32s
  platformio_options:
    board_build.mcu: esp32s2
    platform_packages: framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#idf-release/v4.2
```
See issues:
* https://github.com/esphome/feature-requests/issues/871
* https://github.com/esphome/feature-requests/issues/996

## Description:

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by misery** [comment]
Of course this is not compatible with old version. I don't see something like ```#if VERSION >= 2.0``` or something like this. How can I add this without breaking current stuff?

My ESP32-S2 works now with this patch and that YAML.

**Comment by misery** [comment]
See https://github.com/espressif/arduino-esp32/pull/4485

**Comment by brentru** [comment]
@misery  This PR works perfectly on my ESP32-S2, an Adafruit Metro ESP32-S2.

**Comment by makermelissa** [comment]
I tested this out as well with success.

**Comment by brentru** [comment]
@misery Initialization of the I2C component on esp32-s2 platforms is failing.

**Comment by misery** [comment]
I did not tested it. But I think that should be another pull request for S2.

**Comment by makermelissa** [comment]
If you put logging into Verbose mode, a some additional items fail to compile in wifi_component_esp32.cpp

**Comment by luizribeiro** [comment]
I may have bandwidth to help here one of these weekends. What is missing here before we can have this merged?

**Comment by misery** [comment]
The branch `idf-release/v4.2` from `espressif/arduino-esp32` was closed. The new `idf-master` works with this PR, too.

It works https://github.com/espressif/arduino-esp32/pull/4996

File: esphome/components/logger/logger.cpp [hunk 1]
```diff
@@ -136,7 +136,11 @@ void Logger::pre_setup() {
         break;
 #ifdef ARDUINO_ARCH_ESP32
       case UART_SELECTION_UART2:
+#if ESP_IDF_VERSION_MAJOR >= 4
+        this->hw_serial_ = &Serial1;
```
Comment (OttoWinter) [inline esphome/components/logger/logger.cpp:140]: Sorry for the late review. I'm wondering about this line... Does `Serial2` no longer exist for newer IDF versions? or why is this needed? Has `Serial2` for some reason been renamed and does the new interface point still point to where the USB controller is usually attached?

**Review by OttoWinter — COMMENTED** [review]
LGTM! thanks!

I have one question about the change for logger though, please see the review comment.

File: esphome/components/logger/logger.cpp (hunk 1 above)
Comment (misery) [reply esphome/components/logger/logger.cpp:140]: Well, I can attach to my ESP32-S2 with USB without a problem.

Looks like they disabled it with the initial commit for ESP32-S2: https://github.com/espressif/arduino-esp32/commit/0a262244e637bd46769cb1a25f2d9b3bc9ca18ef#diff-ba8df4ae3e43ea0d34307f26d6def560d7c458d88bf6fb2ddd315a94a60dbdaeR118

`CONFIG_IDF_TARGET_ESP32` is not enabled for ESP32-S2 as it would be `CONFIG_IDF_TARGET_ESP32S2`.
```
#if CONFIG_IDF_TARGET_ESP32
extern HardwareSerial Serial2;
#endif
```

```
src/esphome/components/logger/logger.cpp: In member function 'void esphome::logger::Logger::pre_setup()':
src/esphome/components/logger/logger.cpp:142:29: error: 'Serial2' was not declared in this scope
         this->hw_serial_ = &Serial2;
                             ^~~~~~~
src/esphome/components/logger/logger.cpp:142:29: note: suggested alternative: 'Serial1'
         this->hw_serial_ = &Serial2;
                             ^~~~~~~
                             Serial1

```

Comment (misery) [reply esphome/components/logger/logger.cpp:140]: Should we switch to this?

```
#if CONFIG_IDF_TARGET_ESP32S2
        this->hw_serial_ = &Serial1;
#else
        this->hw_serial_ = &Serial2;
#endif
```

Comment (misery) [reply esphome/components/logger/logger.cpp:140]: @OttoWinter Would be nice if we could finalize for next release. 🙂

**Comment by OttoWinter** [comment]
Just checked out the branch with this config (the idf4.2 branch was merged, so using master instead):

```yaml
esphome:
  platformio_options:
    board_build.mcu: esp32s2
    platform_packages: framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#master
```

However, there are some new build issues now

```
In file included from src/esphome/components/wifi/wifi_component_esp32.cpp:16:
src/esphome/components/wifi/wifi_component_esp32.cpp: In member function 'void esphome::wifi::WiFiComponent::wifi_event_callback_(arduino_event_id_t, arduino_event_info_t)':                     
src/esphome/components/wifi/wifi_component_esp32.cpp:430:81: error: invalid initialization of reference of type 'const ip4_addr_t&' {aka 'const ip4_addr&'} from expression of type 'esp_ip4_addr_t' {aka 'esp_ip4_addr'}                                                                          
       ESP_LOGV(TAG, "Event: Got IP static_ip=%s gateway=%s", format_ip4_addr(it.ip).c_str(),
                                                                              ~~~^~              
src/esphome/core/log.h:79:91: note: in definition of macro 'esph_log_v'                          
   esp_log_printf_(ESPHOME_LOG_LEVEL_VERBOSE, tag, __LINE__, ESPHOME_LOG_FORMAT(format), ##__VA_ARGS__)
...
```

I'll look into what interface has changed.

**Comment by makermelissa** [comment]
Just a head's up, there's a related issue that you may run into once this is in place. The issue is that currently WiFi does not connect on the ESP32-S2 when I2C is in use. We believe we have traced it to a bug in esp-idf (PR submitted: https://github.com/espressif/esp-idf/pull/6848). esp-idf (along with the bug) sits inside of arduino-esp32 repo in the `tools/sdk/esp32s2/lib/` folder as precompiled code, so it's not so easy to fix.

**Comment by misery** [comment]
> Just a head's up, there's a related issue that you may run into once this is in place. The issue is that currently WiFi does not connect on the ESP32-S2 when I2C is in use. We believe we have traced it to a bug in esp-idf (PR submitted: [espressif/esp-idf#6848](https://github.com/espressif/esp-idf/pull/6848)). esp-idf (along with the bug) sits inside of arduino-esp32 repo in the `tools/sdk/esp32s2/lib/` folder as precompiled code, so it's not so easy to fix.
> 
> This issue isn't in esphome itself and it doesn't directly affect this particular PR. I just wanted you to be aware.

Thanks for information! But this shoudn't be a blocker for this PR since it's a bug in esp-idf itself. :-)

**Review by jesserockz — APPROVED** [review]
LGTM, I just made these same changes the other day when playing around with the versions.

**Comment by mmakaay** [comment]
Ha! I was playing with these as well, just two days ago.
GMTA!  (sneakily qualifying myself as "GM")

Thanks and kudos! 👍

**Comment by kvn1351** [comment]
I got the following error:

`A fatal error occurred: Segment loaded at 0x3f000120 lands in same 64KB flash mapping as segment loaded at 0x3f000020. Can't generate binary. Suggest changing linker script or ELF to merge sections.
*** [/data/test/.pioenvs/test/firmware.bin] Error 2
========================== [FAILED] Took 2.99 seconds ==========================`

So I had to change from baster to the alpha like this:

`platform_packages: framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#2.0.0-alpha1` 

Thanks a lot to @jesserockz for coming up with this.

**Comment by misery** [comment]
@kvn1351 I tried current state today. I works for me.... could you retry? Maybe they fixed it in current master branch again.

**Comment by makermelissa** [comment]
I'm getting the same results as @kvn1351 even now.
